File: mapir_camera_ros2/src/camera_node.cpp
#include <algorithm>
#include <cctype>
#include <chrono>
#include <cmath>
#include <memory>
#include <string>

#include "camera_info_manager/camera_info_manager.hpp"
#include "cv_bridge/cv_bridge.hpp"
#include "mapir_camera_cpp/gstreamer_pipeline.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/image.hpp"

#include <opencv2/opencv.hpp>

namespace {

cv::VideoCapture open_v4l2_capture(const std::string &video_device) {
  if (!video_device.empty() && std::all_of(video_device.begin(), video_device.end(), ::isdigit)) {
    return cv::VideoCapture(std::stoi(video_device), cv::CAP_V4L2);
  }
  if (video_device.rfind("/dev/video", 0) == 0) {
    try {
      int index = std::stoi(video_device.substr(10));
      return cv::VideoCapture(index, cv::CAP_V4L2);
    } catch (const std::exception &) {
    }
  }
  return cv::VideoCapture(video_device, cv::CAP_V4L2);
}

}  // namespace

namespace mapir_camera_cpp {

class MapirCameraCppNode : public rclcpp::Node {
 public:
  MapirCameraCppNode()
  : Node("mapir_camera_cpp") {
    declare_parameter("debug", false);
    declare_parameter("debug_period_s", 1.0);
    declare_parameter("video_device", "/dev/video0");
    declare_parameter("image_width", 1280);
    declare_parameter("image_height", 720);
    declare_parameter("framerate", 30.0);
    declare_parameter("frame_id", "mapir3_optical_frame");
    declare_parameter("camera_name", "mapir3_ocn");
    declare_parameter("camera_info_url", "");
    declare_parameter("pixel_format", "MJPG");
    declare_parameter("use_gstreamer", false);
    declare_parameter("gstreamer_pipeline", "");
    declare_parameter("qos_depth", 5);
    declare_parameter("qos_best_effort", true);

    debug_ = get_parameter("debug").as_bool();
    debug_period_s_ = get_parameter("debug_period_s").as_double();

    video_device_ = get_parameter("video_device").as_string();
    req_width_ = get_parameter("image_width").as_int();
    req_height_ = get_parameter("image_height").as_int();
    req_fps_ = get_parameter("framerate").as_double();
    frame_id_ = get_parameter("frame_id").as_string();
    camera_name_ = get_parameter("camera_name").as_string();
    camera_info_url_ = get_parameter("camera_info_url").as_string();
    pixel_format_ = get_parameter("pixel_format").as_string();
    use_gstreamer_ = get_parameter("use_gstreamer").as_bool();
    gstreamer_pipeline_ = get_parameter("gstreamer_pipeline").as_string();

    qos_depth_ = std::max(1, static_cast<int>(get_parameter("qos_depth").as_int()));
    qos_best_effort_ = get_parameter("qos_best_effort").as_bool();

    if (req_fps_ <= 0.0) {
      RCLCPP_WARN(get_logger(), "Invalid framerate; defaulting to 30 Hz");
      req_fps_ = 30.0;
    }

    if (debug_) {
      RCLCPP_INFO(get_logger(), "Params: device=%s size=%dx%d fps=%.2f fmt=%s frame_id=%s camera_name=%s",
                  video_device_.c_str(), req_width_, req_height_, req_fps_, pixel_format_.c_str(),
                  frame_id_.c_str(), camera_name_.c_str());
      RCLCPP_INFO(get_logger(), "QoS: reliability=%s depth=%d",
                  qos_best_effort_ ? "BEST_EFFORT" : "RELIABLE", qos_depth_);
    }
    rclcpp::QoS qos_profile(qos_depth_);
    qos_profile.reliability(qos_best_effort_ ? RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT
                                             : RMW_QOS_POLICY_RELIABILITY_RELIABLE);
    qos_profile.durability(RMW_QOS_POLICY_DURABILITY_VOLATILE);

    image_pub_ = create_publisher<sensor_msgs::msg::Image>("image_raw", qos_profile);
    cinfo_pub_ = create_publisher<sensor_msgs::msg::CameraInfo>("camera_info", qos_profile);

    cinfo_manager_ = std::make_unique<camera_info_manager::CameraInfoManager>(
      this, camera_name_, camera_info_url_);
    if (!camera_info_url_.empty()) {
      RCLCPP_INFO(get_logger(), "camera calibration URL: %s", camera_info_url_.c_str());
      cinfo_manager_->loadCameraInfo(camera_info_url_);
    }
    if (debug_ && !(cinfo_manager_ && cinfo_manager_->isCalibrated())) {
      RCLCPP_WARN(get_logger(), "No valid calibration loaded; using default CameraInfo.");
    }

    open_camera();

    last_stats_t_ = std::chrono::steady_clock::now();
    last_fail_log_t_ = last_stats_t_;
    last_frame_t_ = last_stats_t_;

    timer_period_ = std::chrono::duration<double>(1.0 / req_fps_);
    timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(timer_period_),
      std::bind(&MapirCameraCppNode::capture_and_publish, this));

    RCLCPP_INFO(get_logger(), "MAPIR Survey3 Camera (C++) started: %dx%d @ %.1f Hz (%s)",
                width_, height_, req_fps_, video_device_.c_str());
  }

  ~MapirCameraCppNode() override {
    if (cap_.isOpened()) {
      cap_.release();
    }
  }

 private:
  void open_camera() {
    std::string fmt = pixel_format_;
    for (auto &c : fmt) {
      c = static_cast<char>(::toupper(c));
    }

    if (use_gstreamer_) {
      GstreamerConfig cfg;
      cfg.device = video_device_;
      cfg.width = req_width_;
      cfg.height = req_height_;
      cfg.fps = static_cast<int>(std::round(req_fps_));
      cfg.pixel_format = fmt;
      cfg.custom_pipeline = gstreamer_pipeline_;
      std::string pipeline = BuildGstreamerPipeline(cfg);
      if (debug_) {
        RCLCPP_INFO(get_logger(), "GStreamer pipeline: %s", pipeline.c_str());
      }
      cap_.open(pipeline, cv::CAP_GSTREAMER);
      width_ = req_width_;
      height_ = req_height_;
    } else {
      cap_ = open_v4l2_capture(video_device_);
      if (!cap_.isOpened()) {
        throw std::runtime_error("Failed to open video device");
      }
      if (fmt != "MJPG" && fmt != "H264") {
        RCLCPP_WARN(get_logger(), "Unsupported pixel_format '%s', forcing MJPG", fmt.c_str());
        fmt = "MJPG";
      }
      int fourcc = cv::VideoWriter::fourcc(fmt[0], fmt[1], fmt[2], fmt[3]);
      cap_.set(cv::CAP_PROP_FOURCC, fourcc);
      cap_.set(cv::CAP_PROP_FRAME_WIDTH, req_width_);
      cap_.set(cv::CAP_PROP_FRAME_HEIGHT, req_height_);
      cap_.set(cv::CAP_PROP_FPS, req_fps_);

      width_ = static_cast<int>(cap_.get(cv::CAP_PROP_FRAME_WIDTH));
      height_ = static_cast<int>(cap_.get(cv::CAP_PROP_FRAME_HEIGHT));
    }

    if (!cap_.isOpened()) {
      throw std::runtime_error("Camera open failed");
    }

    cv::Mat test_frame;
    if (cap_.read(test_frame) && !test_frame.empty()) {
      width_ = test_frame.cols;
      height_ = test_frame.rows;
      RCLCPP_INFO(get_logger(), "First frame OK: shape=%dx%d", width_, height_);
    } else {
      RCLCPP_WARN(get_logger(), "Camera opened but no frames could be read");
    }
  }

  sensor_msgs::msg::CameraInfo default_camerainfo() const {
    sensor_msgs::msg::CameraInfo info;
    info.width = width_;
    info.height = height_;
    return info;
  }

  void capture_and_publish() {
    cv::Mat frame;
    if (!cap_.read(frame) || frame.empty()) {
      fail_reads_ += 1;
      if (debug_) {
        auto now = std::chrono::steady_clock::now();
        const double dt_s = std::chrono::duration<double>(now - last_fail_log_t_).count();
        if (dt_s >= debug_period_s_) {
          RCLCPP_WARN(get_logger(), "Camera read failed (fail_reads=%d)", fail_reads_);
          last_fail_log_t_ = now;
        }
      }
      return;
    }

    auto stamp = get_clock()->now();
    auto now = std::chrono::steady_clock::now();
    if (last_frame_t_.time_since_epoch().count() > 0) {
      const double dt_s = std::chrono::duration<double>(now - last_frame_t_).count();
      if (dt_s > max_inter_frame_s_) {
        max_inter_frame_s_ = dt_s;
      }
    }
    last_frame_t_ = now;

    std_msgs::msg::Header header;
    header.stamp = stamp;
    header.frame_id = frame_id_;

    auto img_msg = cv_bridge::CvImage(header, "bgr8", frame).toImageMsg();

    sensor_msgs::msg::CameraInfo cinfo;
    if (cinfo_manager_ && cinfo_manager_->isCalibrated()) {
      cinfo = cinfo_manager_->getCameraInfo();
    } else {
      cinfo = default_camerainfo();
    }
    cinfo.header = header;

    image_pub_->publish(*img_msg);
    cinfo_pub_->publish(cinfo);
    pub_frames_ += 1;

    if (debug_) {
      const double stats_dt_s = std::chrono::duration<double>(now - last_stats_t_).count();
      if (stats_dt_s >= debug_period_s_) {
        const int frames_since = pub_frames_ - last_stats_pub_frames_;
        const double est_hz = frames_since / std::max(1e-6, stats_dt_s);
        RCLCPP_INFO(get_logger(),
                    "Stats: published_frames=%d fail_reads=%d est_pub_hz=%.2f max_inter_frame_dt=%.4fs",
                    pub_frames_, fail_reads_, est_hz, max_inter_frame_s_);
        last_stats_pub_frames_ = pub_frames_;
        last_stats_t_ = now;
        max_inter_frame_s_ = 0.0;
      }
    }
  }

  bool debug_ = false;
  double debug_period_s_ = 1.0;

  std::string video_device_;
  int req_width_ = 1280;
  int req_height_ = 720;
  double req_fps_ = 30.0;
  std::string frame_id_;
  std::string camera_name_;
  std::string camera_info_url_;
  std::string pixel_format_;
  bool use_gstreamer_ = false;
  std::string gstreamer_pipeline_;

  int qos_depth_ = 5;
  bool qos_best_effort_ = true;

  int width_ = 0;
  int height_ = 0;
  int fail_reads_ = 0;
  int pub_frames_ = 0;
  int last_stats_pub_frames_ = 0;
  double max_inter_frame_s_ = 0.0;
  std::chrono::steady_clock::time_point last_stats_t_;
  std::chrono::steady_clock::time_point last_fail_log_t_;
  std::chrono::steady_clock::time_point last_frame_t_;

  cv::VideoCapture cap_;

  rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
  rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr cinfo_pub_;
  std::unique_ptr<camera_info_manager::CameraInfoManager> cinfo_manager_;

  std::chrono::duration<double> timer_period_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace mapir_camera_cpp

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<mapir_camera_cpp::MapirCameraCppNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: mapir_camera_ros2/camera_node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Author: Duda Andrada
# Maintainer: Duda Andrada <duda.andrada@isr.uc.pt>
# License: GNU General Public License v3.0 (GPL-3.0-only)
# Repository: mapir_survey3
#
# Description:
#   ROS 2 Jazzy camera driver for MAPIR Survey3 cameras.
#   Publishes /<ns>/image_raw and /<ns>/camera_info at up to 60 Hz.
#   Uses OpenCV with V4L2 backend and forces MJPG/H264 (device dependent).
#   Uses CameraInfoManager; publishes default CameraInfo if calibration missing.
#   Includes best-practice debug logging controlled by a 'debug' parameter.
#
# Notes:
#   - Best-practice QoS for camera streams is BEST_EFFORT (RViz/rqt compatible).
#   - Debug logs are throttled with debug_period_s to avoid console spam at 30–60 Hz.
#

from __future__ import annotations

import threading
import time
from typing import Optional

from camera_info_manager import CameraInfoManager
from cv_bridge import CvBridge
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from sensor_msgs.msg import CameraInfo, Image
from std_msgs.msg import String

from mapir_camera_ros2.camera_capture import (
    apply_runtime_vignette_correction,
    configure_camera,
    configure_vignette_if_enabled,
    get_camerainfo_safe,
    open_camera,
    update_dimensions_from_frame,
)
from mapir_camera_ros2.camera_controls import apply_uvc_controls_if_requested
from mapir_camera_ros2.camera_metadata import (
    build_frame_metadata,
    publish_metadata,
    shutdown_metadata,
    start_metadata_reader,
)
from mapir_camera_ros2.camera_params import (
    apply_camera_parameters,
    build_publisher_qos,
    declare_camera_parameters,
    load_camera_parameters,
    log_parameter_banner,
)


class MapirSurvey3CameraNode(Node):
    """
    MAPIR Survey3 camera driver node.

    Publishes (under the node namespace):
      - image_raw   (sensor_msgs/Image)
      - camera_info (sensor_msgs/CameraInfo)

    Debugging:
      - debug:=true enables additional logging.
      - debug_period_s throttles periodic debug output.
    """

    def __init__(self) -> None:
        super().__init__('mapir_camera')

        declare_camera_parameters(self)
        params = load_camera_parameters(self)
        apply_camera_parameters(self, params)

        if self.req_fps <= 0.0:
            self.get_logger().warn('Invalid framerate; defaulting to 30 Hz')
            self.req_fps = 30.0

        # -----------------------
        # Debug banner
        # -----------------------
        if self.debug:
            log_parameter_banner(self, params)

        # -----------------------
        # QoS (best practice for camera images is BEST_EFFORT)
        # -----------------------
        self.pub_qos = build_publisher_qos(params)

        if self.debug:
            self.get_logger().info(
                f'Publisher QoS: reliability={self.pub_qos.reliability.name}, '
                f'depth={self.pub_qos.depth}'
            )

        # -----------------------
        # Open camera (force V4L2)
        # -----------------------
        self.cap = open_camera(self, self.video_device)
        if not self.cap.isOpened():
            self.get_logger().error(f'Failed to open video device: {self.video_device}')
            raise RuntimeError('Camera open failed')

        configure_camera(self)

        # Validate first frame
        ok, test_frame = self.cap.read()
        if not ok or test_frame is None:
            self.get_logger().error(
                'Camera opened but no frames could be read. '
                'Check /dev/video permissions, device busy, or negotiated pixel format.'
            )
        else:
            update_dimensions_from_frame(self, test_frame)
            self.get_logger().info(
                f'First frame OK: shape={test_frame.shape}, dtype={test_frame.dtype}'
            )

        self.bridge = CvBridge()

        # -----------------------
        # Publishers
        # -----------------------
        self.image_pub = self.create_publisher(Image, 'image_raw', self.pub_qos)
        self.cinfo_pub = self.create_publisher(CameraInfo, 'camera_info', self.pub_qos)
        self.metadata_pub: Optional[rclpy.publisher.Publisher] = None
        if self.metadata_enabled:
            self.metadata_pub = self.create_publisher(String, self.metadata_topic, self.pub_qos)

        # -----------------------
        # CameraInfoManager
        # -----------------------
        self.cinfo_manager = CameraInfoManager(
            node=self,
            cname=self.camera_name,
            url=self.camera_info_url,
        )

        loaded = False
        if self.camera_info_url:
            self.get_logger().info(f'camera calibration URL: {self.camera_info_url}')

        try:
            loaded = self.cinfo_manager.loadCameraInfo()
        except Exception as ex:
            self.get_logger().warn(f'CameraInfo loadCameraInfo() raised: {ex}')

        if self.camera_info_url and loaded:
            self.get_logger().info(f'Loaded camera calibration: {self.camera_info_url}')
        else:
            self.get_logger().warn(
                'No valid calibration loaded; publishing default (uncalibrated) CameraInfo. '
                'Set "camera_info_url" to a file:// YAML to enable calibration.'
            )

        # -----------------------
        # Runtime stats (debug) — separate throttles
        # -----------------------
        self._fail_reads = 0
        self._pub_frames = 0

        now = time.time()
        self._last_stats_log_t = now
        self._last_fail_log_t = 0.0
        self._last_stats_pub_frames = 0

        self._last_frame_t: Optional[float] = None
        self._max_dt = 0.0
        self._metadata_lock = threading.Lock()
        self._latest_metadata: dict = {}
        self._metadata_thread: Optional[threading.Thread] = None
        self._stop_metadata = threading.Event()
        self._metadata_proc: Optional[subprocess.Popen] = None
        self._metadata_bytes = 0
        self._metadata_records = 0
        self._metadata_log_file: Optional[object] = None
        self._metadata_log_records_since_flush = 0
        self._flat_fields_bgr: Optional[tuple] = None
        self._uvc_controls_device_applied = ''
        self._uvc_controls_requested: dict[str, int] = {}
        self._uvc_controls_applied: dict[str, str] = {}
        self._uvc_controls_locked = False

        apply_uvc_controls_if_requested(self)

        if self.metadata_enabled:
            start_metadata_reader(self)

        configure_vignette_if_enabled(self)

        # Timer
        self.timer_period = 1.0 / self.req_fps
        self.timer = self.create_timer(self.timer_period, self.capture_and_publish)

        self.get_logger().info(
            f'MAPIR Survey3 Camera started: {self.req_width}x{self.req_height} '
            f'@ {self.req_fps:.1f} Hz ({self.video_device}), '
            f'fmt={self.pixel_format}, qos={self.pub_qos.reliability.name}'
        )

    def capture_and_publish(self) -> None:
        """Timer callback: capture one frame and publish Image + CameraInfo."""
        ret, frame = self.cap.read()

        # Read failure (throttled independently)
        if not ret or frame is None:
            self._fail_reads += 1
            if self.debug:
                now = time.time()
                if (now - self._last_fail_log_t) >= self.debug_period_s:
                    self.get_logger().warn(
                        f'Camera read failed (fail_reads={self._fail_reads}) '
                        f'(device={self.video_device}, fmt={self.pixel_format})'
                    )
                    self._last_fail_log_t = now
            return

        if self.vignette_enabled and self._flat_fields_bgr is not None:
            frame = apply_runtime_vignette_correction(self, frame)

        # Timing stats
        now_t = time.time()
        if self._last_frame_t is not None:
            dt = now_t - self._last_frame_t
            if dt > self._max_dt:
                self._max_dt = dt
        self._last_frame_t = now_t

        stamp = self.get_clock().now().to_msg()

        img_msg = self.bridge.cv2_to_imgmsg(frame, encoding='bgr8')
        img_msg.header.stamp = stamp
        img_msg.header.frame_id = self.frame_id

        cinfo_msg = get_camerainfo_safe(self)
        cinfo_msg.header.stamp = stamp
        cinfo_msg.header.frame_id = self.frame_id

        self.image_pub.publish(img_msg)
        self.cinfo_pub.publish(cinfo_msg)
        self._pub_frames += 1
        if self.metadata_enabled:
            payload = build_frame_metadata(self, stamp, img_msg)
            publish_metadata(self, payload)

        # Periodic debug stats (throttled)
        if self.debug:
            now = time.time()
            if (now - self._last_stats_log_t) >= self.debug_period_s:
                frames_since = self._pub_frames - self._last_stats_pub_frames
                dt = max(1e-6, now - self._last_stats_log_t)
                est_hz = frames_since / dt

                self.get_logger().info(
                    'Stats: '
                    f'published_frames={self._pub_frames}, '
                    f'fail_reads={self._fail_reads}, '
                    f'est_pub_hz={est_hz:.2f}, '
                    f'timer_period={self.timer_period:.4f}s, '
                    f'max_inter_frame_dt={self._max_dt:.4f}s'
                )

                self._last_stats_pub_frames = self._pub_frames
                self._last_stats_log_t = now

    def destroy_node(self):
        """Ensure camera resource is released cleanly."""
        try:
            shutdown_metadata(self)
            if hasattr(self, 'cap') and self.cap is not None:
                self.cap.release()
        finally:
            super().destroy_node()


def main(args=None) -> None:
    rclpy.init(args=args)
    node = MapirSurvey3CameraNode()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()
